x/leverage/keeper/msg_server_test.go

```diff
@@ -661,6 +661,134 @@ func (s *IntegrationTestSuite) TestMsgMaxWithdraw() {
+
+	// create and fund another supplier with 200 UMEE
+	other := s.newAccount(coin.New(umeeDenom, 200_000000))
+	s.supply(other, coin.New(umeeDenom, 200_000000))
+
+	// create a borrower with 2000 ATOM, then collateralize 2000 of supplied ATOM
+	// borrow 750 UMEE
+	borrower := s.newAccount(coin.New(atomDenom, 2000_000000))
+	s.supply(borrower, coin.New(atomDenom, 2000_000000))
+	s.collateralize(borrower, coin.New("u/"+atomDenom, 2000_000000))
+	s.borrow(borrower, coin.New(umeeDenom, 750_000000))
+
+	// the other user executes MaxWithdraw
+	msg := &types.MsgMaxWithdraw{
+		Supplier: other.String(),
+		Denom:    umeeDenom,
+	}
+
+	// expected UMEE withdraw amount (90 from available_module_liquidity and 0 from available_module_collateral):
+	// 		= liquidity - min_collateral_liquidity * collateral
+	// 		= 250 - 0.2*800
+	//		= 90
+	// since available_module_liquidity is 90 and it's less than the users_spendable_tokens
+	// only the available_module_liquidity can be withdrawn
+
+	// verify the outputs of withdraw function
+	resp, err := srv.MaxWithdraw(ctx, msg)
+	require.NoError(err)
+	require.Equal(coin.New(umeeDenom, 90_000000), resp.Received)
+}
+
+func (s *IntegrationTestSuite) TestMsgMaxWithdraw_NoAvailableModuleLiquidity() {
+	app, ctx, srv, require := s.app, s.ctx, s.msgSrvr, s.Require()
+
+	// overriding UMEE token settings, changing MinCollateralLiquidity to 0.2
+	umeeToken := newToken(umeeDenom, "UMEE", 6)
+	umeeToken.MinCollateralLiquidity = sdk.MustNewDecFromStr("0.2")
+	require.NoError(app.LeverageKeeper.SetTokenSettings(ctx, umeeToken))
+
+	// create and fund a supplier with 800 UMEE, then collateralize 800 of supplied UMEE
+	supplier := s.newAccount(coin.New(umeeDenom, 800_000000))
+	s.supply(supplier, coin.New(umeeDenom, 800_000000))
+	s.collateralize(supplier, coin.New("u/"+umeeDenom, 800_000000))
+
+	// create and fund another supplier with 200 UMEE
+	other := s.newAccount(coin.New(umeeDenom, 200_000000))
+	s.supply(other, coin.New(umeeDenom, 200_000000))
+	s.collateralize(other, coin.New("u/"+umeeDenom, 200_000000))
+
+	// create a borrower with 2000 ATOM, then collateralize 2000 of supplied ATOM
+	// borrow 750 UMEE
+	borrower := s.newAccount(coin.New(atomDenom, 2000_000000))
+	s.supply(borrower, coin.New(atomDenom, 2000_000000))
+	s.collateralize(borrower, coin.New("u/"+atomDenom, 2000_000000))
+	s.borrow(borrower, coin.New(umeeDenom, 800_000000))
+
+	// the other user executes MaxWithdraw
+	msg := &types.MsgMaxWithdraw{
+		Supplier: other.String(),
+		Denom:    umeeDenom,
+	}
+
+	// expected UMEE withdraw amount (0 from available_module_liquidity and 0 from available_module_collateral):
+	// 		= liquidity - min_collateral_liquidity * collateral
+	// 		= 200 - 0.2*1000
+	//		= 0
+	// since available_module_liquidity is zero, nothing to withdraw
+
+	// verify the outputs of withdraw function
+	resp, err := srv.MaxWithdraw(ctx, msg)
+	require.NoError(err)
+	require.Equal(coin.New(umeeDenom, 0_000000), resp.Received)
+}
+
```

Comment:
Might be good to add a complex test, where:
- uToken exchange rate > 1
- amount from liquidity and amount from collateral are _both_ nonzero

Reply:
done